### PR TITLE
[Speed Up] `BaseDataElement` supports `pin_memory` 

### PR DESCRIPTION
**Pull request by hhaAndroid**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Custom object `BaseDataElement` supports `pin_memory` function to speed up training. 

Note: The underlying logic has been modified, it's best to check if the relevant repo has been speeded up.

## Modification

- BaseDataElement add `pin_memory` function
- Add UT

## BC-breaking (Optional)
None

## Use cases (Optional)

dataloader needs to be enabled `pin_memory=True`.

### CODE REVIEW

File: mmengine/structures/base_data_element.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 import torch
+from torch.utils.data._utils.pin_memory import pin_memory
```
Comment (RangiLyu) [inline mmengine/structures/base_data_element.py:8]: it is not safe to call a private method

File: mmengine/structures/base_data_element.py [hunk 2]
```diff
@@ -561,6 +562,16 @@ def to_dict(self) -> dict:
             for k, v in self.all_items()
         }
 
+    def pin_memory(self) -> 'BaseDataElement':
+        """BaseDataElement supports pin_memory function to speed up
```
Comment (RangiLyu) [inline mmengine/structures/base_data_element.py:566]: need to indicate what is pin_memory in the docstring

File: mmengine/structures/base_data_element.py [hunk 3]
```diff
@@ -561,6 +562,16 @@ def to_dict(self) -> dict:
...
+        training."""
+        for k, v in self.items():
+            if not isinstance(v, BaseDataElement):
```
Comment (RangiLyu) [inline mmengine/structures/base_data_element.py:569]: can simply replace with if hasattr(v, 'pin_memory') 
just follow the `cuda()` and `to()` method

File: mmengine/structures/base_data_element.py [hunk 4]
```diff
@@ -561,6 +562,16 @@ def to_dict(self) -> dict:
             for k, v in self.all_items()
         }
 
+    def pin_memory(self) -> 'BaseDataElement':
```
Comment (RangiLyu) [inline mmengine/structures/base_data_element.py:565]: Need to provide `is_pinned` method

File: mmengine/structures/base_data_element.py (hunk 1 above)
Comment (hhaAndroid) [reply mmengine/structures/base_data_element.py:8]: If you can't call it, we can only rewrite it?

File: mmengine/structures/base_data_element.py (hunk 4 above)
Comment (hhaAndroid) [reply mmengine/structures/base_data_element.py:565]: There are many properties inside this object, so if one is not pinned, return false?

File: mmengine/structures/base_data_element.py (hunk 3 above)
Comment (hhaAndroid) [reply mmengine/structures/base_data_element.py:569]: ok

File: mmengine/structures/base_data_element.py (hunk 2 above)
Comment (HAOCHENYE) [reply mmengine/structures/base_data_element.py:566]: If we use the internal function `pin_memory`, seems this if-else is unnecessary. `pin_memory` will handle  various types of inputs.

File: mmengine/structures/base_data_element.py [hunk 5]
```diff
@@ -561,6 +562,16 @@ def to_dict(self) -> dict:
...
+        """BaseDataElement supports pin_memory function to speed up
+        training."""
+        for k, v in self.items():
+            if not isinstance(v, BaseDataElement):
+                setattr(self, k, pin_memory(v))
```
Comment (HAOCHENYE) [inline mmengine/structures/base_data_element.py:570]: Suggest using `self.set_field`

**Pull request closed without merging**